test/macros.jl

```diff
@@ -176,6 +183,75 @@ function build_constraint_keyword_test(ModelType::Type{<:JuMP.AbstractModel})
     end
 end
 
+struct MyConstrType end
+struct BadPosArg end
+function JuMP.build_constraint(
+    _error::Function,
+    f::AffExpr,
+    set::MOI.EqualTo,
+    extra::Type{MyConstrType};
+    d = 0,
+)
+    new_set = MOI.LessThan{Float64}(set.value + d)
+    return JuMP.build_constraint(_error, f, new_set)
+end
+function build_constraint_extra_arg_test(ModelType::Type{<:JuMP.AbstractModel})
+    @testset "build_constraint with extra positional arguments" begin
+        model = ModelType()
+        @variable(model, x)
+        cref = @constraint(model, x == 0, MyConstrType)
+        @test JuMP.constraint_object(cref).set isa MOI.EqualTo{Float64}
+        cref = @constraint(model, c1, x == 0, MyConstrType, d = 1)
+        @test JuMP.constraint_object(cref).set == MOI.LessThan{Float64}(1)
+        @test_macro_throws ErrorException @constraint(model, x == 0, BadPosArg)
+        @test_macro_throws ErrorException @constraint(
+            model,
+            x == 0,
+            BadPosArg,
+            d = 1
+        )
+    end
+end
+
+struct MyInfo
+    var::JuMP.VariableRef
+    value::Float64
+end
+function JuMP.build_constraint(
+    _error::Function,
+    func::AffExpr,
+    set::MOI.AbstractScalarSet,
+    info::MyInfo,
+)
+    func.terms[info.var] *= info.value
+    return JuMP.build_constraint(_error, func, set)
+end
+function JuMP.parse_extra_constraint_args(
```

Comment:
Yeah, this is type-piracy. Please remove.

Reply:
Done